firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/LiveSession.kt

```diff
@@ -0,0 +1,304 @@
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+/**
+ * Represents a live WebSocket session capable of streaming bidirectional content to and from the
+ * server.
+ */
+public class LiveSession
+internal constructor(
+  private val session: ClientWebSocketSession?,
+  private var isRecording: Boolean,
+  private var audioHelper: AudioHelper? = null
+) {
+
+  private val audioQueue = ConcurrentLinkedQueue<ByteArray>()
+  private val playBackQueue = ConcurrentLinkedQueue<ByteArray>()
+  private var startedReceiving = false
+  private var receiveChannel: Channel<Frame> = Channel()
+  private var functionCallChannel: Channel<List<FunctionCallPart>> = Channel()
+
+  @Serializable
+  internal data class ClientContent(
+    @SerialName("turns") val turns: List<Content.Internal>,
+    @SerialName("turn_complete") val turnComplete: Boolean
+  )
+
+  @Serializable
+  internal data class ClientContentSetup(
+    @SerialName("client_content") val clientContent: ClientContent
+  )
+
+  @Serializable internal data class ToolResponseSetup(val toolResponse: ToolResponse)
+  @Serializable
+  internal data class ToolResponse(
+    val functionResponses: List<FunctionResponsePart.Internal.FunctionResponse>
+  )
+
+  @Serializable
+  internal data class ServerContentSetup(
+    @SerialName("serverContent") val serverContent: ServerContent
+  )
+
+  @Serializable
+  internal data class ServerContent(@SerialName("modelTurn") val modelTurn: Content.Internal)
+
+  @Serializable internal data class MediaChunks(val mediaChunks: List<MediaData.Internal>)
+
+  @Serializable internal data class MediaStreamingSetup(val realtimeInput: MediaChunks)
+
+  @Serializable internal data class ToolCallSetup(val toolCall: ToolCall)
+
+  @Serializable
+  internal data class ToolCall(val functionCalls: List<FunctionCallPart.Internal.FunctionCall>)
+
+  /**
+   * Receives all function call responses from the server for the audio conversation feature..
+   *
+   * @return A [Flow] which will emit list of [FunctionCallPart] as they are returned by the model.
+   */
+  public fun receiveAudioConvoFunctionCalls(): Flow<List<FunctionCallPart>> {
+    return functionCallChannel.receiveAsFlow()
+  }
+
+  /**
+   * Starts an audio conversation with the Gemini server, which can only be stopped using
+   * stopAudioConversation.
+   */
+  public suspend fun startAudioConversation() {
+    if (isRecording) {
+      return
+    }
+    functionCallChannel = Channel()
+    isRecording = true
+    audioHelper = AudioHelper()
+    audioHelper!!.setupAudioTrack()
+    val scope = CoroutineScope(Dispatchers.Default)
+    CoroutineScope(Dispatchers.IO).launch {
+      audioHelper!!.startRecording().collect {
+        if (!isRecording) {
+          cancel()
+        }
+        println(it)
+        audioQueue.add(it)
+      }
+    }
+
+    scope.launch {
+      val minBufferSize =
```

Comment:
some of these blocks could be broken out into private functions for readability. Everything in this scope.launch is a good example